distributed/diagnostics/memory_sampler.py

```diff
@@ -0,0 +1,215 @@
+        client: Client, optional
+            client used to connect to the scheduler.
+            Default: use the global client
+        measure: str, optional
+            One of the measures from :class:`distributed.scheduler.MemoryState`.
+            Default: sample process memory
+        interval: float, optional
+            sampling interval, in seconds.
+            Default: 0.5
+        """
+        if not client:
+            from ..client import get_client
+
+            client = get_client()
+
+        if client.asynchronous:
+            return self._sample_async(label, client, measure, interval)
+        else:
+            return self._sample_sync(label, client, measure, interval)
+
+    @contextmanager
+    def _sample_sync(
+        self, label: str | None, client: Client, measure: str, interval: float
+    ):
+        key = client.sync(
+            client.scheduler.memory_sampler_start,
+            client=client.id,
+            measure=measure,
+            interval=interval,
+        )
+        try:
+            yield
+        finally:
+            samples = client.sync(client.scheduler.memory_sampler_stop, key=key)
+            self.samples[label or key] = samples
+
+    @asynccontextmanager
+    async def _sample_async(
+        self, label: str | None, client: Client, measure: str, interval: float
+    ):
+        key = await client.scheduler.memory_sampler_start(
+            client=client.id, measure=measure, interval=interval
+        )
+        try:
+            yield
+        finally:
+            samples = await client.scheduler.memory_sampler_stop(key=key)
+            self.samples[label or key] = samples
+
+    def to_pandas(self, *, align: bool = False):
+        """Return the data series as a pandas.Dataframe.
+
+        Parameters
+        ==========
+        align : bool, optional
+            If True, change the absolute timestamps into time deltas from the first
+            sample of each series, so that different series can be visualized side by
+            side. If False (the default), use absolute timestamps.
+        """
+        import pandas as pd
+
+        ss = {}
+        for (label, s_list) in self.samples.items():
+            assert s_list  # There's always at least one sasmple
+            s = pd.DataFrame(s_list).set_index(0)[1]
+            s.index = pd.to_datetime(s.index, unit="s")
+            s.name = label
+            if align:
+                # convert datetime to timedelta from the first sample
+                s.index -= s.index[0]
+            ss[label] = s
+
+        df = pd.DataFrame(ss)
+
+        if len(ss) > 1:
+            # Forward-fill NaNs in the middle of a series created either by overlapping
+            # sampling time range or by align=True. Do not ffill series beyond their
+            # last sample.
+            df = df.ffill().where(~pd.isna(df.bfill()))
+
+        return df
+
+    def plot(self, *, align: bool = False, **kwargs):
```

Comment:
I'm not sure I see much value to having a bespoke `plot` function that is essentially identical to `ms.to_pandas().plot()`, especially since it also has an implicit matplotlib dependency.

Reply:
it relieves the user from figuring out what the axes and units are

Reply:
I also like the convenience of it for that reason